sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala

```diff
@@ -0,0 +1,194 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream}
+
+import com.google.common.io.ByteStreams
+
+import org.apache.spark.storage.StorageLevel
+import org.apache.spark.unsafe.Platform
+import org.apache.spark.unsafe.array.ByteArrayMethods
+import org.apache.spark.unsafe.memory.{TaskMemoryManager, MemoryBlock}
+import org.apache.spark.{TaskContext, SparkEnv}
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.io.CompressionCodec
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.UnsafeRow
+
+/**
+ * ::Experimental::
+ *
+ * @since 1.6.0
+ */
+@Experimental
+object TungstenCache {
+  def cache(
+      child: SparkPlan,
+      compressionType: Option[String] = None,
+      blockSize: Int = 4 * 1000 * 1000): RDD[InternalRow] = {
+    val numFields = child.schema.length
+    val childRDD: RDD[UnsafeRow] = {
+      if (child.outputsUnsafeRows) {
+        child.execute().asInstanceOf[RDD[UnsafeRow]]
+      } else {
+        ConvertToUnsafe(child).execute().asInstanceOf[RDD[UnsafeRow]]
+      }
+    }
+    val cachedRDD: RDD[MemoryBlock] = childRDD.mapPartitions { rowIterator =>
+      // Note: this buffering is used to let us hold onto rows when rolling across block boundaries.
+      val bufferedRowIterator = rowIterator.buffered
+      val taskMemoryManager = TaskContext.get().taskMemoryManager()
+      val compressionCodec: Option[CompressionCodec] =
+        compressionType.map { t => CompressionCodec.createCodec(SparkEnv.get.conf, t) }
+      new Iterator[MemoryBlock] {
+        // NOTE: This assumes that size of every row < blockSize
+        def next(): MemoryBlock = {
+          // Packs rows into a `blockSize` bytes contiguous block of memory, starting a new block
+          // whenever the current fills up.
+          // Each row is laid out in memory as [rowSize (int)|rowData (rowSize bytes)]
+          val block = taskMemoryManager.allocateUnchecked(blockSize)
+
+          var currOffset = 0
+          while (bufferedRowIterator.hasNext && currOffset < blockSize) {
+            val currRow = bufferedRowIterator.head
+            val recordSize = 4 + currRow.getSizeInBytes
+            if (currOffset + recordSize < blockSize) {
+              Platform.putInt(
+                block.getBaseObject, block.getBaseOffset + currOffset, currRow.getSizeInBytes)
+              currRow.writeToMemory(block.getBaseObject, block.getBaseOffset + currOffset + 4)
+              bufferedRowIterator.next()
+            }
```

Comment:
@JoshRosen @yhuai adding the -1 terminator and the above code fixes the extra row bug on dogfood